src/main/java/org/jenkinsci/remoting/engine/JnlpProtocol.java

```diff
@@ -49,16 +57,44 @@ protected JnlpProtocol(String secretKey, String slaveName) {
     public abstract String getName();
 
     /**
+     * Performs a handshake with the master and establishes a {@link Channel}.
+     *
+     * @param socket The established {@link Socket} connection to the master.
+     * @param channelBuilder The {@link ChannelBuilder} to use.
+     * @return The established channel if successful, else null.
+     * @throws IOException
+     */
+    @Nullable
+    public Channel establishChannel(Socket socket, ChannelBuilder channelBuilder) throws IOException {
+        DataOutputStream outputStream = new DataOutputStream(socket.getOutputStream());
+        BufferedInputStream inputStream = new BufferedInputStream(socket.getInputStream());
+        if(performHandshake(outputStream, inputStream)) {
+            return buildChannel(socket, channelBuilder);
```

Comment:
I'm guessing this pull request introduced a bug? If so apologies - could you provide a link to the bug or more context?

I think I see what you're saying, here a buffered input stream is created for the socket which is then not used to create the channel, which means some bytes read in the original buffered input stream could be lost - is that correct?